Exhaustive check for listening processes

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'checks'

**Context**: 

#3092 
#264 

**Summary**:

Main:
- Add a pre-check in MetalK8s bootstrap and restore about all listening ports
- Add a CI tests to ensure the "MetalK8s ports pre-checks" list is exhaustive
- Document all listening processes by MetalK8s

Side:
- tox: set tox default basepython to `python3.6`
- bootstrap, restore, expand: set up MetalK8s grains before call "checking local node"
- disable ssh host key checking in salt-master container
- explicitly install `psutil` python module and include it in thin by default
- ci: always untaint bootstrap node in "single-node"
- give pillar values directly in `deploy_node` orchestrate for `metalk8s_checks.node` in salt-ssh

**Acceptance criteria**:

Bootstrap:
```
$ nc -l 4505 & /srv/scality/metalk8s-2.8.0-dev/bootstrap.sh 
[2] 16717
> Determine the OS... done [0s]
> Checking that BootstrapConfiguration is present... done [0s]
> Pre-minion system tests... done [0s]
> Configure internal repositories... done [1s]
> Check mandatory packages presence... done [5s]
> Disabling Salt minion service... done [1s]
> Stopping Salt minion service... done [0s]
> Installing mandatory packages... done [13s]
> Configuring Salt minion to run in local mode... done [6s]
> Ensure archive is available... done [2s]
> Calculating Salt grains in local mode... done [2s]
> Checking CA minion... done [3s]
> Checking local node... fail [3s]

Failure while running step 'Checking local node'

Command: check_local_node

Output:

<< BEGIN >>
Error running 'metalk8s_checks.node': Node bootstrap: Salt master publisher should be listening on 1.2.3.4:4505 but nc(16717) listen on 0.0.0.0.
<< END >>

This script will now exit
```

Expansion:

```
$ salt-run state.sls metalk8s.orchestrate.deploy_node saltenv=metalk8s-2.8.0-dev pillar="{'orchestrate': {'node_name': 'node-1'}}"
[...]
bootstrap_master:
[...]
----------
          ID: Check node
    Function: metalk8s.saltutil_cmd
        Name: metalk8s_checks.node
      Result: False
     Comment: Running function metalk8s_checks.node failed on minions: node-1
     Started: 15:20:45.515555
    Duration: 3299.884 ms
     Changes:   
              ----------
              node-1:
                  ----------
                  retcode:
                      1
                  stderr:
                      Error running 'metalk8s_checks.node': Node node-1: Calico node should be listening on 127.0.0.1:9099 but nc(18868) listen on 0.0.0.0.
                  stdout:

Summary for bootstrap_master
------------
Succeeded: 3 (changed=4)
Failed:    1
------------
Total states run:     3
Total run time:  62.025 s

```

---

Fixes: #3092 
Fixes: #264

TeddyAndrieux committed:
salt: Add a module function to check available Ports

Add a module that use `metalk8s_network.get_listening_processes` to
check listening connection on the local node and check that running
process match what is expected by MetalK8s

Sees: #3092

TeddyAndrieux committed:
tox: Force basepython to `python3.6` for unit-tests

Tox entry for unit-tests does not work properly with newer python
version so force the python to python3.6

TeddyAndrieux committed:
salt: Add all listening ports as part of listening process check

Check for all MetalK8s listening processes in `metalk8s_checks.ports`
function

Fixes: #3092

TeddyAndrieux committed:
salt: Add a module function to list listening process

Add a simple salt function that use `psutil` to list all listening
processes on every ports

Sees: #3092

TeddyAndrieux committed:
tests: Add a check for exhaustive list of listening processes

Add a check in CI to make sure the listening processes list we check in
MetalK8s is exhaustif with everything running on the platform.

NOTE: This check is supposed to run on specific environment that only
have a small list of listening processes other than MetalK8s ones

Sees: #3092

TeddyAndrieux committed:
script,salt: Move set of MetalK8s grains earlier

We need the network grain to be set when checking local node ports, so
move this grains set in the common.sh script and call it in both
restore and bootstrap script, just before checking local node.
Also on expansion add a salt state ssh to set grains on node before
checking the node

TeddyAndrieux committed:
eve: Always "untaint" bootstrap in single node

It does not make sense to have tainted bootstrap when running a single
node cluster, so by default in the CI always untaint the bootstrap node

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-pre-checks-for-ports` into
`development/2.8`
* [w/2.9/improvement/add-pre-checks-for-ports](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/improvement/add-pre-checks-for-ports) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

NicolasT commented on salt/_modules/metalk8s_network.py:
```diff
@@ -3,6 +3,10 @@
 import itertools
 import logging
 
+# Note: psutil is a dependency of Salt RPMs in MetalK8s context we
```
Would it make sense to still explicitly install the package(s) our custom modules/states/.... depend on when we install the Salt package in some script(s), even when they're dependencies of said Salt packages and would hence be installed implicitly?

TeddyAndrieux replied:
Hmm sure we can, make sense I will add it

TeddyAndrieux replied:
Done

bert-e commented:
# History mismatch



Merge commit #a95eed6f4d2a438775ae62d86bbac7ba147bfddc on the integration branch
`w/2.9/improvement/add-pre-checks-for-ports` is merging a branch which is neither the current
branch `improvement/add-pre-checks-for-ports` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/add-pre-checks-for-ports` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux committed:
salt: Always include `psutil` module in the thin

When checking ports with Salt SSH we need psutil module, so always
include this module in the Salt SSH thin archive

TeddyAndrieux committed:
script,salt: Explicitly install `python36-psutil`

We need psutil python module to run some command as part of our MetalK8s
custom Salt execution module, nowaday `python36-psutil` is a
dependency of `salt` but it may not be the case in the futur, so
explicitly install this package in bootstrap, restore script and during
expansion

TeddyAndrieux committed:
images/salt-master: Disable ssh host key checking

Salt-ssh do not properly handle ssh host key ignoring even if
`no_host_keys` is set to True in salt master configuration file.

Sees: https://github.com/saltstack/salt/issues/59691

To avoid this issue just disable host key checking for ssh in the
salt-master container directly

bert-e commented:
# History mismatch



Merge commit #7a269e47f147c8071eaec3cbf7ef878df2f844c2 on the integration branch
`w/2.9/improvement/add-pre-checks-for-ports` is merging a branch which is neither the current
branch `improvement/add-pre-checks-for-ports` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/add-pre-checks-for-ports` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux committed:
salt: Use `service_cidr` pillar for `metalk8s_checks` when using salt-ssh

When running execution module with salt-ssh we cannot access pillar
content due to a bug in Salt.
So give the pillar value `service_cidr` from salt-master when executing
the `metalk8s_checks.node` with salt-ssh

Sees: https://github.com/saltstack/salt/issues/28503

TeddyAndrieux committed:
docs: List listening processes of MetalK8s

List all listening processes by MetalK8s so that we have the full list
of port that need to be free on every MetalK8s hosts.
This list is directly "computed" from `salt/defaults.yaml` so that if a
port change or get added the doc get updated automatically

Fixes: #264

TeddyAndrieux committed:
changelog: Add changelog entry for listening processes check

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-pre-checks-for-ports` into
`development/2.8`
* [w/2.9/improvement/add-pre-checks-for-ports](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/improvement/add-pre-checks-for-ports) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

bert-e commented:
# History mismatch



Merge commit #a78fe4fa474ac7c22764da5e90d7b33b57012cb7 on the integration branch
`w/2.9/improvement/add-pre-checks-for-ports` is merging a branch which is neither the current
branch `improvement/add-pre-checks-for-ports` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/add-pre-checks-for-ports` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-pre-checks-for-ports` into
`development/2.8`
* [w/2.9/improvement/add-pre-checks-for-ports](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/improvement/add-pre-checks-for-ports) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

bert-e commented:
# History mismatch



Merge commit #c5584090ff25ac4a59b3761bc668373730445f57 on the integration branch
`w/2.9/improvement/add-pre-checks-for-ports` is merging a branch which is neither the current
branch `improvement/add-pre-checks-for-ports` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/add-pre-checks-for-ports` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-pre-checks-for-ports` into
`development/2.8`
* [w/2.9/improvement/add-pre-checks-for-ports](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/improvement/add-pre-checks-for-ports) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

gdemonet commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -20,6 +20,26 @@ Install python36:
     - raw_shell: true
     - roster: kubernetes
 
+Set grains ssh:
+  salt.state:
+    - ssh: true
+    - roster: kubernetes
+    - tgt: {{ node_name }}
+    - saltenv: metalk8s-{{ version }}
+    - sls:
+      - metalk8s.node.grains
+    - require:
+      - metalk8s: Install python36
```
Do we still need the plain "Set grains" after this, then?

TeddyAndrieux replied:
Yes for several reasons:
- grains are not set in the same place when running ssh and salt-minion (basically when we run this state in salt-ssh it will just put grains the `thin_dir` of the "minion" and not in the `/etc/salt/grains` file as the salt-minion will)
- this `Set grains ssh` is only run if the salt-minion is not yet configured (and we may need one day to add some grains and need to run this state also during upgrade ...)

gdemonet commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -20,6 +20,26 @@ Install python36:
+    - require:
+      - metalk8s: Install python36
+
+Refresh grains:
+  metalk8s.saltutil_cmd:
+    - name: saltutil.sync_grains
+    - tgt: {{ node_name }}
+    - ssh: true
+    - roster: kubernetes
+    - require:
+      - salt: Set grains ssh
```
Isn't this used to sync custom grain modules? If so, wouldn't it make sense to move before the "Set grains ssh" (though I'm not sure it uses custom grain modules)?

TeddyAndrieux replied:
Normally it's used to sync module yes but it also refreshes the grains and btw I don't really know why but just calling a `refresh_grains` does not work in salt-ssh here.

Btw we need to do it here otherwise we do not see the metalk8s grains in salt-ssh.
(Also in MetalK8s we do not have any custom grain module)

gdemonet commented on scripts/bootstrap.sh.in:
```diff
@@ -64,14 +64,7 @@ declare -a PACKAGES=(
-            saltenv=metalk8s-@@VERSION \
-            pillarenv=metalk8s-@@VERSION
+
 
```
(nit) Not sure we need these empty lines :)

TeddyAndrieux replied:
Right

gdemonet commented on scripts/bootstrap.sh.in:
```diff
@@ -57,6 +57,7 @@ SALT_CALL=${SALT_CALL:-salt-call}
 
 declare -a PACKAGES=(
     salt-minion-@@SALT_VERSION
+    python3-psutil
```
(this may be just me not understanding what we do) Don't we need to use the "correct" package name depending on the target OS? I see we have `override` names in `buildchain/versions.py`, but not sure how these are impacting installation at runtime :confused: (maybe it's missing some comments :sweat_smile:)

TeddyAndrieux replied:
No, it's just today in bootstrap script we do not have any "override name" for packages we install, and btw `python3-psutil` work for both RHEL7 and RHEL8. It will install `python36-psutil` on RHEL7 and `python3-psutil` on RHEL8.

I could have use some more rendering (same as what we do for `@@VERSION`, `@@SALT_VERSION`) just did not want to add more complexity here just for a `psutil` package that will be installed anyway by salt-minion  But I could if it's confusing, or at least add a comment for the moment

TeddyAndrieux replied:
I added a comment I think it's good enough for this specific case we do not need complex rendering stuff

gdemonet replied:
Agreed. I'm thinking it would make sense, once we rework our scripts to be part of a single Python-based CLI, to get this list of packages directly from `versions.py`, and filter based on OS at runtime.

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+
+    Arguments:
+        listening_process (dict): override the list of ports expected to be
+            unused (or bind to a MetalK8s process).
```
(nit) `s/bind/bound/`

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+        listening_process (dict): override the list of ports expected to be
+            unused (or bind to a MetalK8s process).
+        raises (bool): the method will raise if there is any unexpected process
+            bind on a MetalK8s port.
```
(nit) `s/bind/bound/`

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+    {
+        '<address>': {
+            'expected': '<processes>',
+            'descritpion': '<description>',
```
(typo) `s/descritpion/description/`

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+        '<address>': {
+            'expected': '<processes>',
+            'descritpion': '<description>',
+            'mandatory': 'True/False'
```
Expected as a string?

TeddyAndrieux replied:
No, you right

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+    for address, matching in listening_process.items():
+        if isinstance(address, int):
+            address = str(address)
+        ip = None
+        port = address
+        if len(address.rsplit(":")) == 2:
+            ip, port = address.rsplit(":")
```
Note that you can use `rpartition` to write this in one line:

```python
ip, _, port = address.rpartition(':')
```

<details>
<summary>Small example:</summary>

```python
>>> addresses = ["127.0.0.1:6443", "control_plane_ip:8443", "4507", ":3128", "more:colons:1234"]
>>> for addr in addresses:
...     ip, _, port = addr.rpartition(":")
...     print(f"Address: {addr:20}\tIP: {ip:16}\tPort: {port}")
... 
Address: 127.0.0.1:6443      	IP: 127.0.0.1       	Port: 6443
Address: control_plane_ip:8443	IP: control_plane_ip	Port: 8443
Address: 4507                	IP:                 	Port: 4507
Address: :3128               	IP:                 	Port: 3128
Address: more:colons:1234    	IP: more:colons     	Port: 1234
```

</details>

TeddyAndrieux replied:
Right

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+            processes.append(None)
+
+        process_on_port = all_listen_connections.get(
+            str(port), all_listen_connections.get(int(port))
```
(nit) Can't we make sure the result of `__salt__["metalk8s_network.get_listening_processes"]()` is only indexed with strings?

TeddyAndrieux replied:
Hmm you right we control this module so it's easy to do. I will

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+
+        success = False
+        error_process = {}
+        # If one of this process is expected one, succeed
```
(nit) `s/this process/these processes/`

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+        success = False
+        error_process = {}
+        # If one of this process is expected one, succeed
+        for process in processes:
```
Took me some time to understand this loop :confused:, maybe a comment could help others.

If I understand correctly, this loop tries to find the first successful match, among the `processes` list.

Say we have:

```python
listening_process = {
  "127.0.0.1:6443": {
    "expected": "kube-apiserver",
    "mandatory": False
  }
}
```

In this case, we will have:

```python
processes = ["kube-apiserver", None]
```

So if `kube-apiserver` is listening, we will not hit the `None` option, and everything will go fine.
In essence, setting `"mandatory": False` means having a fallback to "nothing listening" after having tried all `"expected"` patterns.

TeddyAndrieux replied:
Yes exactly but note also that if you would have put the None manually and got a 
```python
processes = [None, "kube-apiserver"]
```
The result would have been the same, just error for None if apiserver listening but this error is ignored in the resulting message anyway since one of the process match.

I will add a comment

gdemonet commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+                "nothing listening"
+                if not error_process
+                else " and ".join(
+                    "{}({}) listen on {}".format(proc["name"], proc["pid"], proc_ip)
```
I'd suggest the following, to improve (slightly) the clarity of such messages:

```python
"'{p[name]}' (PID: {p[pid]}) listens on {addr}".format(p=proc, addr=proc_ip)
```

Most important part in this suggestion is `s/listen/listens/`

gdemonet commented on tests/post/steps/test_network.py:
```diff
@@ -0,0 +1,108 @@
+    nodes = k8s_client.list_node()
+
+    if len(nodes.items) != 1:
+        pytest.skip("We skip single node cluster for this test")
```
I think you meant "skip multi nodes clusters" :wink:

TeddyAndrieux replied:
:smiley: right

gdemonet commented on tests/post/steps/test_network.py:
```diff
@@ -0,0 +1,108 @@
+            if ip == "127.0.0.1" and process["name"] == "containerd":
+                continue
+
+            assert False, "Process {}({}) listen on {}:{}".format(
+                process["name"], process["pid"], ip, port
+            )
```
Would make sense to collect all unexpected processes in a single error to output, so if someone added more than one, the test would tell them :)

TeddyAndrieux replied:
I was thinking of it as well but do not have the motivation the time I finished this PR, but ... ok I will :smile:

gdemonet commented on CHANGELOG.md:
```diff
@@ -44,6 +44,10 @@
   (PR [#3152](https://github.com/scality/metalk8s/pull/3152))
 
+- [#3092](https://github.com/scality/metalk8s/issues/3092) - Check if all needed
+  addresses are free, or already used by a MetalK8s, process
```
`s/MetalK8s, process/MetalK8s process/`

gdemonet commented on docs/operation/listening_processes.rst:
```diff
@@ -0,0 +1,29 @@
+Listening processes
```
Title-case in our docs is requiring us to always capitalize words (unless they are for liaison): `Listening Processes` in this case.

gdemonet commented on docs/operation/listening_processes.rst:
```diff
@@ -0,0 +1,29 @@
+with each other, sometimes locally, sometimes between machines in the cluster,
+or with the end user.
+
+Several addresses must be free for MetalK8s processes depending on the
+node roles.
```
```rst
Depending on their :ref:`roles <node-roles>`, nodes must have
several addresses available for MetalK8s processes to bind.
```

gdemonet commented on docs/operation/listening_processes.rst:
```diff
@@ -0,0 +1,29 @@
+   {% if role == "node" %}
+     {%- set role = "all" %}
+   {%- endif %}
+   Listening processes on {{ role }} nodes
```
Title case: `Listening Processes on {{ role | capitalize }} Nodes`

gdemonet commented on docs/operation/listening_processes.rst:
```diff
@@ -0,0 +1,29 @@
+   Listening processes on {{ role }} nodes
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+   .. csv-table::
+      :header: "Address", "Description"
+      :align: left
```
To make things neat, I'd suggest applying constant width to all tables. Maybe adding `:width: 100%` would be enough, in this case.

TeddyAndrieux replied:
Sure, I will test

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #da1d9ace3fd00584bdcb39f7739bafdd62a1f561 on the integration branch
`w/2.9/improvement/add-pre-checks-for-ports` is merging a branch which is neither the current
branch `improvement/add-pre-checks-for-ports` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/add-pre-checks-for-ports` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-pre-checks-for-ports` into
`development/2.8`
* [w/2.9/improvement/add-pre-checks-for-ports](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/improvement/add-pre-checks-for-ports) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

TeddyAndrieux commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

gdemonet reviewed (approved):
:shipit:

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
